Fix invalid rounding value for TimeIntervalRounding close to DST transitions

cbuescher opened the pull request:
There are edge cases where rounding a date to a certain interval using a time
zone with DST shifts can currently cause the rounded date to be bigger than the
original date. This happens when rounding a date closely after a DST start and
the rounded date falls into the DST gap.

Here is an example for CET time zone, where local time is set forward by one
hour at 2016-03-27T02:00:00+01:00 to 2016-03-27T03:00:00.000+02:00:

The date 2016-03-27T03:01:00.000+02:00 (1459040460000) which is just after the
DST change is first converted to local time (1459047660000). If we then apply
interval rounding for a 14m interval in local time, this  takes us to
1459047240000, which unfortunately falls into the DST gap.  When converting
this back to UTC, joda provides options to throw exceptions on illegal dates
like this, or correct this by adjusting the date to the new time zone offset.
We currently do the later, but this leads to converting this illegal date back
to 2016-03-27T03:54:00.000+02:00 (1459043640000), giving us a date that is
larger than the original date we wanted to round.

This change fixes this by using the "strict" option of 'convertLocalToUTC()'
to detect rounded dates that fall into the DST gap. If this happens, we can use
the time of the DST change instead as the interval start.

Even before this change, intervals around DST shifts like this can be shorter
than the desired interval.  This, for example, happens when the requested
interval width doesn't completely fit into the remaining time span when the DST
shift happens. For example, using a 14m interval in UTC+1 (CET before DST
starts) leads to the following valid rounding values around the time where DST
happens:

2016-03-27T01:30:00+01:00
2016-03-27T01:44:00+01:00
2016-03-27T01:58:00+01:00
2016-03-27T02:12:00+01:00
2016-03-27T02:26:00+01:00
...

while the rounding values in UTC+2 (CET after DST start) are placed like this
around the same time:

2016-03-27T02:40:00+02:00
2016-03-27T02:54:00+02:00
2016-03-27T03:08:00+02:00
2016-03-27T03:22:00+02:00
...

From this we can see then when we switch from UTC+1 to UTC+2 at 02:00 the last
rounding value in UTC+1 is at 01:58 and the first valid one in UTC+2 is at
03:08, so even if we decide to put all the dates in between into one rounding
interval, it will only cover 10 minutes. With this change we choose to use the
moment of DST shift as an aditional interval separator, leaving us with a 2min
interval from [01:58,02:00) before the shift and an 8min interval from
[03:00,03:08) after the shift.

This change also adds tests for the above example and adds randomization to the
existing TimeIntervalRounding tests.

cbuescher committed:
Fix invalid rounding value for TimeIntervalRounding close to DST transition

There are edge cases where rounding a date to a certain interval using a time
zone with DST shifts can currently cause the rounded date to be bigger than the
original date. This happens when rounding a date closely after a DST start and
the rounded date falls into the DST gap.

Here is an example for CET time zone, where local time is set forward by one
hour at 2016-03-27T02:00:00+01:00 to 2016-03-27T03:00:00.000+02:00:

The date 2016-03-27T03:01:00.000+02:00 (1459040460000) which is just after the
DST change is first converted to local time (1459047660000). If we then apply
interval rounding for a 14m interval in local time, this  takes us to
1459047240000, which unfortunately falls into the DST gap.  When converting
this back to UTC, joda provides options to throw exceptions on illegal dates
like this, or correct this by adjusting the date to the new time zone offset.
We currently do the later, but this leads to converting this illegal date back
to 2016-03-27T03:54:00.000+02:00 (1459043640000), giving us a date that is
larger than the original date we wanted to round.

This change fixes this by using the "strict" option of 'convertLocalToUTC()'
to detect rounded dates that fall into the DST gap. If this happens, we can use
the time of the DST change instead as the interval start.

Even before this change, intervals around DST shifts like this can be shorter
than the desired interval.  This, for example, happens when the requested
interval width doesn't completely fit into the remaining time span when the DST
shift happens. For example, using a 14m interval in UTC+1 (CET before DST
starts) leads to the following valid rounding values around the time where DST
happens:

2016-03-27T01:30:00+01:00
2016-03-27T01:44:00+01:00
2016-03-27T01:58:00+01:00
2016-03-27T02:12:00+01:00
2016-03-27T02:26:00+01:00
...

while the rounding values in UTC+2 (CET after DST start) are placed like this
around the same time:

2016-03-27T02:40:00+02:00
2016-03-27T02:54:00+02:00
2016-03-27T03:08:00+02:00
2016-03-27T03:22:00+02:00
...

From this we can see then when we switch from UTC+1 to UTC+2 at 02:00 the last
rounding value in UTC+1 is at 01:58 and the first valid one in UTC+2 is at
03:08, so even if we decide to put all the dates in between into one rounding
interval, it will only cover 10 minutes. With this change we choose to use the
moment of DST shift as an aditional interval separator, leaving us with a 2min
interval from [01:58,02:00) before the shift and an 8min interval from
[03:00,03:08) after the shift.

This change also adds tests for the above example and adds randomization to the
existing TimeIntervalRounding tests.

cbuescher commented on core/src/main/java/org/elasticsearch/common/rounding/TimeZoneRounding.java:
```diff
@@ -218,7 +219,19 @@ public byte id() {
-            return timeZone.convertLocalToUTC(rounded, false, utcMillis);
+            try {
+                return timeZone.convertLocalToUTC(rounded, true, utcMillis);
+            } catch (IllegalInstantException e) {
```
Note: I don't really like to detect this special case by exception, but the only other way I see how this can be detected is copy the underlying Joda code here or try to use a heuristic like (timeZone.convertLocalToUTC(rounded) > utcMillis) to detect that the rounded value falls into the DST gap.

jimferenczi replied:
Maybe DateTimeZone.isLocalDateTimeGap ?

cbuescher replied:
@jimferenczi thanks, I looked into that method. The problem is that it takes a LocalDateTime object which we would have to construct each time from the local time in millis, and I saw that the way it determines whether the date is in a gap is also by try-catching an IllegalInstantException. So performance-wise unfortunately this doesn't buy us much, but we would need to construct another object on top of that. So I'd like to leave it as-is.

jimferenczi replied:
I agree this is ok as is.

cbuescher commented on core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java:
```diff
@@ -310,23 +310,53 @@ public void testTimeInterval_Kathmandu_DST_Start() {
+                assertThat("Values smaller than rounded value should round further down", rounding.round(roundedDate - 1),
+                        lessThan(roundedDate));
+
+                if (tz.isFixed()) {
```
The following assumtions are not (yet) all true when using time zones that are not fixed. I still want to randomize the existing test, so we can at least test the previous invariants for the `round()` method for all cases.

jimferenczi commented on core/src/main/java/org/elasticsearch/common/rounding/TimeZoneRounding.java:
```diff
@@ -218,7 +219,19 @@ public byte id() {
+                 * This would violate the invariant that the rounding operation
+                 * should be idempotent.
+                 */
+                return timeZone.previousTransition(utcMillis) + 1;
```
We should document the fact that it will creates buckets smaller than the interval ? Only when using timezone though so we should also add a line about the limitations of using a timezone conversion.

cbuescher replied:
Would something like the documentation change requested in https://github.com/elastic/elasticsearch/issues/18805 be enough? I can also add another comment here but having buckets smaller than the specified interval is possible regardless of this edge case.

jimferenczi commented:
@cbuescher the documentation change in https://github.com/elastic/elasticsearch/issues/18805 is enough.  
LGTM

cbuescher committed:
Don't throw IllegalInstantException to determine DST gap

By taking the logic from DateTimeZone#convertLocalToUTC(long, boolean) we
can avoid throwing the exception.

cbuescher commented:
@jimferenczi thanks for the review, I ended up finding another way to detect the edge case without the try-catch logic which I think is much better and added that change in the last commit. Mind to take another quick look only at that change before I merge?

jimferenczi commented on core/src/main/java/org/elasticsearch/common/rounding/TimeZoneRounding.java:
```diff
@@ -218,7 +219,53 @@ public byte id() {
             long rounded = Rounding.Interval.roundValue(Rounding.Interval.roundKey(timeLocal, interval), interval);
-            return timeZone.convertLocalToUTC(rounded, false, utcMillis);
+            long roundedUTC;
+            if (timeZone.isFixed() || isInDSTGap(rounded, timeZone) == false) {
```
Maybe you can move the timeZone.isFixed test in the isInDSTGap directly ?

cbuescher replied:
good idea, will do.

jimferenczi commented:
@cbuescher left one minor comment.
LGTM

cbuescher commented:
Merged to 2.x branch with 780cfb1.
